File: archer/cli/commands/validate.py
import click
import asyncio
import logging
import textwrap

from typing import Optional

from archer.core import SecretValidator
from archer.core.variables import parse_var_args, format_var_name_for_cli
from archer.log_setup import setup_logging

from archer.templates import TemplateLoader
from archer.constants import (
    MODE_SINGLE,
    MODE_MULTIPART,
    SUCCESS_INDICATOR,
    FAILURE_INDICATOR,
    OPT_VAR,
    MISSING_REQUIRED_VARIABLES,
    UNEXPECTED_VARIABLES,
    SECRET_VALID,
    VALIDATION_SUCCESS,
    VALIDATION_FAILED,
)

logger = logging.getLogger(__name__)


@click.command()
@click.argument('template_name')
@click.argument('secret', required=False)
@click.option('--template-file', type=click.Path(exists=True, file_okay=True, dir_okay=False), help='Load template from specific file')
@click.option('--var', 'var_args', multiple=True, help='Variable in format key=value (for multipart templates)')
@click.option('--verbose', '-v', is_flag=True, help='Enable verbose logging')
@click.option('--debug', '-d', is_flag=True, help='Enable debug logging')
def validate(template_name: str, secret: Optional[str], template_file: Optional[str], var_args: tuple, verbose: bool, debug: bool) -> None:
    """Validate a secret using the specified template.

    Validates API secrets against endpoints based on template configuration.
    Template mode determines the required arguments and usage pattern.
    """
    
    help_text = textwrap.dedent("""\
    \b
    Arguments:
      TEMPLATE_NAME  Name of the template to use for validation
      SECRET         Secret to validate (only for single mode templates)

    \b
    Options:
      --template-file FILE      Load template from specific file instead of built-in
      --var TEXT               Variable in format key=value (for multipart templates only)
      --verbose, -v            Enable verbose logging
      --debug, -d              Enable debug logging
      --help                   Show this message and exit

    \b
    Usage Examples:

    \b
    1. SINGLE MODE TEMPLATES:
       Validate simple API tokens that require only the secret value.
       
       archer validate github ghp_xxxxxxxxxxxxxxxxxxxx
       archer validate openai sk-xxxxxxxxxxxxxxxxxxxxxxxx
       archer validate npm npm_xxxxxxxxxxxxxxxxxxxxxxxx

    \b
    2. MULTIPART MODE TEMPLATES:
       Validate APIs requiring multiple parameters using --var options.
       
       archer validate ghost --var base-url=https://myblog.com --var api-token=xxxxx
       archer validate stripe --var secret-key=sk_test_xxxxx --var publishable-key=pk_test_xxxxx
       archer validate supabase --var project-url=https://xxx.supabase.co --var anon-key=eyJxxx

    \b
    3. CUSTOM TEMPLATE FILES:
       Use your own YAML template files for validation.
       
       archer validate myapi --template-file ./custom-api.yaml sk_xxxxxxxxxxxxx
       archer validate custom --template-file ./multipart.yaml --var token=xxx --var url=https://api.example.com

    \b
    Variable Format Requirements:
      For multipart templates, variables must be provided in kebab-case format:
      
      ✓ Correct:   --var api-token=xxx --var base-url=https://example.com
      ✗ Incorrect: --var apiToken=xxx --var base_url=https://example.com
      
      Use 'archer info <template>' to see required variables for any template.
      Use 'archer list' to see which templates are single vs multipart mode.
    """)
    
    setup_logging(verbose, debug)
    logger.info(f"Starting secret validation process for template '{template_name}'")

    async def _validate() -> None:
        # Validate arguments
        if not template_file and not template_name:
            click.echo(f"{FAILURE_INDICATOR} Either template name or --template-file must be provided.")
            click.echo("Usage: archer validate <template_name> <secret>")
            click.echo("   or: archer validate --template-file <path> <secret>")
            raise click.ClickException("Missing template specification")

        # Load template to determine mode
        if template_file:
            # Load from specific file
            template_loader = TemplateLoader()
            template = template_loader.get_template(template_file)
            validation_identifier = template_file
        else:
            # Default directory
            template_loader = TemplateLoader()
            template = template_loader.get_template(template_name)
            validation_identifier = template_name

        if not template:
            if template_file:
                click.echo(f"{FAILURE_INDICATOR} Template file '{template_file}' not found or invalid.")
            else:
                click.echo(f"{FAILURE_INDICATOR} Template '{template_name}' not found.")
            raise click.ClickException("Template not found")

        # Use the same template loader for validation
        validator = SecretValidator()



        # Handle based on template mode
        if template.mode == MODE_SINGLE:
            if not secret:
                click.echo(f"{FAILURE_INDICATOR} Template '{template.name}' is in single mode. Please provide a secret.")
                if template_file:
                    click.echo(f"Usage: archer validate --template-file {template_file} <secret>")
                else:
                    click.echo(f"Usage: archer validate {template_name} <secret>")
                raise click.ClickException("Secret required for single mode template")

            if var_args:
                click.echo(f"{FAILURE_INDICATOR} Template '{template.name}' is in single mode. Variables are not supported.")
                if template_file:
                    click.echo(f"Usage: archer validate --template-file {template_file} <secret>")
                else:
                    click.echo(f"Usage: archer validate {template_name} <secret>")
                raise click.ClickException("Variables not supported for single mode template")

            result = await validator.validate_secret(validation_identifier, secret)

        elif template.mode == MODE_MULTIPART:
            if secret:
                click.echo(f"{FAILURE_INDICATOR} Template '{template.name}' is in multipart mode. Please use --var arguments.")

                # Show expected variables
                var_examples = []
                if template.required_variables:
                    for var in template.required_variables:
                        cli_name = format_var_name_for_cli(var)
                        var_examples.append(f"{OPT_VAR} {cli_name}=<value>")

                # Show appropriate usage based on template source
                if template_file:
                    click.echo(f"Usage: archer validate --template-file {template_file} {' '.join(var_examples)}")
                else:
                    click.echo(f"Usage: archer validate {template_name} {' '.join(var_examples)}")
                raise click.ClickException("Use --var arguments for multipart template")

            if not var_args:
                click.echo(f"{FAILURE_INDICATOR} Template '{template.name}' requires variables.")

                # Show expected variables
                var_examples = []
                if template.required_variables:
                    for var in template.required_variables:
                        cli_name = format_var_name_for_cli(var)
                        var_examples.append(f"{OPT_VAR} {cli_name}=<value>")

                # Show appropriate usage based on template source
                if template_file:
                    click.echo(f"Usage: archer validate --template-file {template_file} {' '.join(var_examples)}")
                else:
                    click.echo(f"Usage: archer validate {template_name} {' '.join(var_examples)}")
                raise click.ClickException("Variables required for multipart template")

            try:
                # Parse variables
                variables = parse_var_args(list(var_args))

                # Validate all required variables are provided
                missing_vars = []
                if template.required_variables:
                    for required_var in template.required_variables:
                        if required_var not in variables:
                            cli_name = format_var_name_for_cli(required_var)
                            missing_vars.append(cli_name)

                if missing_vars:
                    click.echo(f"{FAILURE_INDICATOR} {MISSING_REQUIRED_VARIABLES.format(vars=', '.join(missing_vars))}")
                    raise click.ClickException("Missing required variables")

                # Check for unexpected variables
                unexpected_vars = []
                if template.required_variables:
                    for provided_var in variables.keys():
                        if provided_var not in template.required_variables:
                            cli_name = format_var_name_for_cli(provided_var)
                            unexpected_vars.append(cli_name)

                if unexpected_vars:
                    click.echo(f"{FAILURE_INDICATOR} {UNEXPECTED_VARIABLES.format(vars=', '.join(unexpected_vars))}")
                    if template.required_variables:
                        expected_vars = [format_var_name_for_cli(var) for var in template.required_variables]
                        click.echo(f"Expected variables: {', '.join(expected_vars)}")
                    raise click.ClickException("Unexpected variables provided")

                result = await validator.validate_secret_multipart(validation_identifier, variables)

            except ValueError as e:
                click.echo(f"{FAILURE_INDICATOR} Variable parsing error: {str(e)}")
                raise click.ClickException("Invalid variable format")
        else:
            click.echo(f"{FAILURE_INDICATOR} Template mode '{template.mode}' is not supported.")
            raise click.ClickException("Unsupported template mode")

        if result.get("valid", False):
            click.echo(f"{SUCCESS_INDICATOR} {result.get('message', SECRET_VALID)}")
            logger.info(VALIDATION_SUCCESS)
        else:
            error_msg = result.get('error', 'Unknown validation error')
            click.echo(f"{FAILURE_INDICATOR} {error_msg}")
            logger.error(VALIDATION_FAILED.format(error=error_msg))
            raise click.ClickException("Secret validation failed")

    asyncio.run(_validate())

File: internal/cli/validate.go
package cli

import (
	"errors"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/theinfosecguy/archer/internal/constants"
	"github.com/theinfosecguy/archer/internal/logger"
	"github.com/theinfosecguy/archer/internal/models"
	"github.com/theinfosecguy/archer/internal/output"
	"github.com/theinfosecguy/archer/internal/templates"
	"github.com/theinfosecguy/archer/internal/validator"
	"github.com/theinfosecguy/archer/internal/variables"
)

var (
	templateFile string
	varArgs      []string
	verbose      bool
	debug        bool
	outputJSON   string
	jsonOnly     bool
)

var validateCmd = &cobra.Command{
	Use:   "validate TEMPLATE_NAME [SECRET]",
	Short: "Validate a secret using the specified template",
	Long: `Validate a secret using the specified template.

Template mode determines the required arguments and usage pattern.

Single mode:
  # Using environment variable (recommended)
  export ARCHER_SECRET="ghp_xxxxxxxxxxxxxxxxxxxx"
  archer validate github

  # Using command-line argument (shows security warning)
  archer validate github ghp_xxxxxxxxxxxxxxxxxxxx

Multipart mode:
  # Using environment variables (recommended)
  export ARCHER_VAR_BASE_URL="https://myblog.com"
  export ARCHER_VAR_API_TOKEN="xxxxx"
  archer validate ghost

  # Using --var flags (shows security warning)
  archer validate ghost --var base-url=https://myblog.com --var api-token=xxxxx

Security:
  Environment variables prevent secrets from appearing in shell history and process lists.`,
	Args: cobra.MinimumNArgs(1),
	RunE: runValidate,
}

func init() {
	validateCmd.Flags().StringVar(&templateFile, "template-file", "", "Load template from specific file")
	validateCmd.Flags().StringArrayVar(&varArgs, "var", []string{}, "Variable in format key=value (for multipart templates)")
	validateCmd.Flags().BoolVarP(&verbose, "verbose", "v", false, "Enable verbose logging")
	validateCmd.Flags().BoolVarP(&debug, "debug", "d", false, "Enable debug logging")
	validateCmd.Flags().StringVarP(&outputJSON, "output-json", "o", "", "Write structured validation result to JSON file")
	validateCmd.Flags().BoolVar(&jsonOnly, "json-only", false, "Suppress normal terminal success output when writing JSON")
}

func runValidate(cmd *cobra.Command, args []string) error {
	// Setup logging based on flags
	if debug {
		logger.SetDebug()
	} else if verbose {
		logger.SetVerbose()
	}

	logger.Info("Starting secret validation process")
	startTime := time.Now().UTC()

	templateName := args[0]
	var secret string
	if len(args) > 1 {
		secret = args[1]
	}

	// Load template to determine mode
	loader := templates.NewTemplateLoader(constants.DefaultTemplatesDir)
	var templateIdentifier string
	if templateFile != "" {
		templateIdentifier = templateFile
	} else {
		templateIdentifier = templateName
	}

	template, err := loader.GetTemplate(templateIdentifier)
	if err != nil {
		errMsg := fmt.Sprintf("Template '%s' not found or invalid", templateIdentifier)
		if outputJSON != "" {
			writeJSONError(outputJSON, templateName, templateFile, nil, nil, startTime, errMsg)
		}
		return fmt.Errorf("%s %s", constants.FailureIndicator, errMsg)
	}

	// Create validator
	v := validator.NewSecretValidator(constants.DefaultTemplatesDir)

	// Validate based on mode
	if template.Mode == constants.ModeSingle {
		return handleSingleMode(v, templateIdentifier, template, secret, startTime)
	} else if template.Mode == constants.ModeMultipart {
		return handleMultipartMode(v, templateIdentifier, template, secret, varArgs, startTime)
	}

	return fmt.Errorf("unsupported template mode: %s", template.Mode)
}

func handleSingleMode(v *validator.SecretValidator, templateIdentifier string, template *models.SecretTemplate, secret string, startTime time.Time) error {
	// Priority 1: Check environment variable
	envSecret := os.Getenv(constants.EnvSecretName)

	var finalSecret string
	var usedCLI bool

	if envSecret != "" {
		finalSecret = envSecret
	} else if secret != "" {
		finalSecret = secret
		usedCLI = true
	} else {
		errMsg := "secret required. Provide via ARCHER_SECRET environment variable or command-line argument"
		if outputJSON != "" {
			vars := map[string]string{constants.SecretVariableName: ""}
			writeJSONError(outputJSON, templateIdentifier, templateFile, template, vars, startTime, errMsg)
		}
		return errors.New(errMsg)
	}

	if len(varArgs) > 0 {
		errMsg := "--var arguments not allowed in single mode"
		if outputJSON != "" {
			vars := map[string]string{constants.SecretVariableName: finalSecret}
			writeJSONError(outputJSON, templateIdentifier, templateFile, template, vars, startTime, errMsg)
		}
		return errors.New(errMsg)
	}

	// Show warning if secret was passed via CLI
	if usedCLI {
		fmt.Fprint(os.Stderr, constants.WarningSecretInCLI)
	}

	// Build variables map for metadata
	vars := map[string]string{constants.SecretVariableName: finalSecret}

	result, err := v.ValidateSecret(templateIdentifier, finalSecret)
	if err != nil {
		if outputJSON != "" {
			writeJSONError(outputJSON, templateIdentifier, templateFile, template, vars, startTime, err.Error())
		}
		return err
	}

	return handleValidationResult(result, template, vars, startTime)
}

func handleMultipartMode(v *validator.SecretValidator, templateIdentifier string, template *models.SecretTemplate, secret string, varArgs []string, startTime time.Time) error {
	if secret != "" {
		errMsg := "secret argument not allowed in multipart mode. Use --var or ARCHER_VAR_* environment variables instead"
		if outputJSON != "" {
			writeJSONError(outputJSON, templateIdentifier, templateFile, template, nil, startTime, errMsg)
		}
		return errors.New(errMsg)
	}

	// Try to get variables from environment first
	envVars := getEnvVariables(template.RequiredVariables)

	var finalVars map[string]string
	var usedCLI bool

	if len(envVars) > 0 {
		// Check if all required variables are present in env
		missingVars := variables.ValidateVariablesProvided(template.RequiredVariables, envVars)
		if len(missingVars) == 0 {
			// All variables available from environment
			finalVars = envVars
		} else if len(varArgs) > 0 {
			// Some missing from env, use CLI args
			parsedVars, err := variables.ParseVarArgs(varArgs)
			if err != nil {
				if outputJSON != "" {
					writeJSONError(outputJSON, templateIdentifier, templateFile, template, envVars, startTime, err.Error())
				}
				return err
			}
			finalVars = parsedVars
			usedCLI = true
		} else {
			errMsg := fmt.Sprintf("missing required variables: %s. Set via ARCHER_VAR_* environment variables or --var flags", strings.Join(missingVars, ", "))
			if outputJSON != "" {
				writeJSONError(outputJSON, templateIdentifier, templateFile, template, envVars, startTime, errMsg)
			}
			return errors.New(errMsg)
		}
	} else if len(varArgs) > 0 {
		// No env vars, use CLI args
		parsedVars, err := variables.ParseVarArgs(varArgs)
		if err != nil {
			if outputJSON != "" {
				writeJSONError(outputJSON, templateIdentifier, templateFile, template, nil, startTime, err.Error())
			}
			return err
		}
		finalVars = parsedVars
		usedCLI = true
	} else {
		errMsg := "--var arguments or ARCHER_VAR_* environment variables required for multipart mode template"
		if outputJSON != "" {
			writeJSONError(outputJSON, templateIdentifier, templateFile, template, nil, startTime, errMsg)
		}
		return errors.New(errMsg)
	}

	// Show warning if variables were passed via CLI
	if usedCLI {
		fmt.Fprint(os.Stderr, constants.WarningSecretInCLI)
	}

	result, err := v.ValidateSecretMultipart(templateIdentifier, finalVars)
	if err != nil {
		if outputJSON != "" {
			writeJSONError(outputJSON, templateIdentifier, templateFile, template, finalVars, startTime, err.Error())
		}
		return err
	}

	return handleValidationResult(result, template, finalVars, startTime)
}

// getEnvVariables retrieves variables from ARCHER_VAR_* environment variables
func getEnvVariables(requiredVariables []string) map[string]string {
	envVars := make(map[string]string)

	for _, varName := range requiredVariables {
		envKey := constants.EnvVarPrefix + varName
		if value := os.Getenv(envKey); value != "" {
			envVars[varName] = value
		}
	}

	return envVars
}

func handleValidationResult(result *models.ValidationResult, template *models.SecretTemplate, vars map[string]string, startTime time.Time) error {
	// Write JSON output if requested
	if outputJSON != "" {
		if err := writeJSONOutput(outputJSON, result, template, vars, startTime); err != nil {
			fmt.Fprintf(os.Stderr, "Failed to write JSON output: %v\n", err)
		}
	}

	// Handle terminal output
	if result.Valid {
		// Only show success message if not in json-only mode
		if !(outputJSON != "" && jsonOnly) {
			fmt.Printf("%s %s\n", constants.SuccessIndicator, result.Message)
		}
		return nil
	}

	// Always show errors even in json-only mode
	return fmt.Errorf("%s %s", constants.FailureIndicator, result.Error)
}

// writeJSONOutput writes successful validation result to JSON file
func writeJSONOutput(filepath string, result *models.ValidationResult, template *models.SecretTemplate, vars map[string]string, startTime time.Time) error {
	endTime := time.Now().UTC()

	// Build masked artifacts
	maskedURL, maskedHeaders := buildMaskedArtifacts(template, vars)

	// Get variable names (without values)
	varNames := make([]string, 0, len(vars))
	for k := range vars {
		varNames = append(varNames, k)
	}

	// Determine source
	source := "builtin"
	if templateFile != "" {
		source = "file"
	}

	// Build request metadata
	requestMeta := models.ValidationRequestMeta{
		Template:             templateFile,
		ResolvedTemplateName: &template.Name,
		Mode:                 &template.Mode,
		Source:               &source,
		Method:               &template.Method,
		APIURLMasked:         &maskedURL,
		HeadersMasked:        maskedHeaders,
		QueryParamsMasked:    nil, // TODO: implement if needed
		VariablesProvided:    varNames,
		StartedAt:            startTime,
		FinishedAt:           endTime,
		DurationMS:           float64(endTime.Sub(startTime).Milliseconds()),
	}

	if templateFile == "" {
		requestMeta.Template = template.Name
	}

	// Build response metadata
	responseMeta := models.ValidationResponseMeta{
		StatusCode:            nil, // TODO: capture from HTTP response
		RequiredFieldsChecked: template.SuccessCriteria.RequiredFields,
		FailedRequiredField:   nil,
		Error:                 nil,
	}

	// Build final JSON structure
	jsonOutput := &models.ValidationResultJSON{
		Command: "validate",
		Version: constants.Version,
		Valid:   result.Valid,
		Request: requestMeta,
		Response: responseMeta,
	}

	if result.Valid {
		jsonOutput.Message = &result.Message
	} else {
		jsonOutput.Error = &result.Error
		responseMeta.Error = &result.Error
	}

	return output.WriteJSONFile(filepath, jsonOutput)
}

// writeJSONError writes error validation result to JSON file
func writeJSONError(filepath string, templateName string, templateFilePath string, template *models.SecretTemplate, vars map[string]string, startTime time.Time, errorMsg string) {
	endTime := time.Now().UTC()

	// Get variable names (without values)
	var varNames []string
	if vars != nil {
		varNames = make([]string, 0, len(vars))
		for k := range vars {
			varNames = append(varNames, k)
		}
	}

	// Build request metadata
	var resolvedName, mode, method *string
	var source *string
	var maskedURL *string
	var maskedHeaders map[string]string

	if template != nil {
		resolvedName = &template.Name
		mode = &template.Mode
		method = &template.Method
		src := "builtin"
		if templateFilePath != "" {
			src = "file"
		}
		source = &src

		// Build masked artifacts if we have vars
		if vars != nil {
			url, headers := buildMaskedArtifacts(template, vars)
			maskedURL = &url
			maskedHeaders = headers
		}
	}

	templateID := templateName
	if templateFilePath != "" {
		templateID = templateFilePath
	}

	requestMeta := models.ValidationRequestMeta{
		Template:             templateID,
		ResolvedTemplateName: resolvedName,
		Mode:                 mode,
		Source:               source,
		Method:               method,
		APIURLMasked:         maskedURL,
		HeadersMasked:        maskedHeaders,
		QueryParamsMasked:    nil,
		VariablesProvided:    varNames,
		StartedAt:            startTime,
		FinishedAt:           endTime,
		DurationMS:           float64(endTime.Sub(startTime).Milliseconds()),
	}

	// Build response metadata
	responseMeta := models.ValidationResponseMeta{
		Error: &errorMsg,
	}

	// Build final JSON structure
	jsonOutput := &models.ValidationResultJSON{
		Command:  "validate",
		Version:  constants.Version,
		Valid:    false,
		Error:    &errorMsg,
		Request:  requestMeta,
		Response: responseMeta,
	}

	// Ignore errors writing JSON error output
	_ = output.WriteJSONFile(filepath, jsonOutput)
}

// buildMaskedArtifacts builds masked URL and headers for JSON output
func buildMaskedArtifacts(template *models.SecretTemplate, vars map[string]string) (string, map[string]string) {
	_, maskedURL := variables.ProcessURL(template.APIURL, vars)
	_, maskedHeaders := variables.ProcessHeaders(template.Request.Headers, vars)
	return maskedURL, maskedHeaders
}
